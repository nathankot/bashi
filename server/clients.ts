import * as r from "redis";
import { Configuration, OpenAIApi } from "openai";

import "https://deno.land/x/dotenv/load.ts";

const whisperEndpoint = Deno.env.get("WHISPER_TRANSCRIBE_ENDPOINT");
if (whisperEndpoint == null || whisperEndpoint.length === 0) {
  throw new Error("WHISPER_TRANSCRIBE_ENDPOINT must be declared");
}

export const redis = r.createClient({
  url: Deno.env.get("REDIS_URL"),
});

export const openai = new OpenAIApi(
  new Configuration({
    apiKey: Deno.env.get("OPENAI_KEY"),
  })
);

export const whisper = {
  async transcribe(audio: ArrayBuffer) {
    const whisperRequest = new FormData();
    whisperRequest.append("audio", new Blob([audio]));
    const whisperResponse = await fetch(whisperEndpoint + "?language=en", {
      method: "POST",
      body: whisperRequest,
    });
    const whisperBody = await whisperResponse.json();
    const text = whisperBody.text;
    if (typeof text !== "string") {
      throw new Error("expected whisper response body to be string");
    }
    return text;
  },
};

export async function setup() {
  console.log("connecting redis client");
  await redis.connect();
}

export async function gracefulStop() {
  console.log("disconnecting redis");
  const rc = await redis;
  await rc.quit();
  return;
}
