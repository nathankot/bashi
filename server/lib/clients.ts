import * as t from "io-ts";
import * as r from "redis";
import { Configuration, OpenAIApi } from "openai";

import "dotenv/load.ts";

const redisUrl = Deno.env.get("REDIS_URL");
if (redisUrl == null || redisUrl.length === 0) {
  throw new Error("REDIS_URL must be declared");
}

export async function withRedis<T>(
  f: (client: ReturnType<typeof r.createClient>) => Promise<T>
): Promise<T> {
  const client = r.createClient({
    url: Deno.env.get("REDIS_URL"),
  });

  let result: T;
  await client.connect();
  try {
    result = await f(client);
  } finally {
    await client.disconnect();
  }

  return result;
}

export const openai = new OpenAIApi(
  new Configuration({
    apiKey: Deno.env.get("OPENAI_KEY"),
  })
);

const whisperEndpoint = Deno.env.get("WHISPER_TRANSCRIBE_ENDPOINT")!;
if (whisperEndpoint == null || whisperEndpoint.length === 0) {
  throw new Error("WHISPER_TRANSCRIBE_ENDPOINT must be declared");
}

export { whisperEndpoint };

const GoogleSearchResponse = t.partial({
  items: t.array(
    t.type({
      title: t.string,
      link: t.string,
      snippet: t.string,
    })
  ),
});

export async function googleSearch(
  q: string,
  signal?: AbortSignal
): Promise<
  {
    link: string;
    title: string;
    snippet: string;
  }[]
> {
  const key = Deno.env.get("GOOGLE_SEARCH_API_KEY");
  const cx = Deno.env.get("GOOGLE_SEARCH_ENGINE_ID");
  if (key == null) {
    throw new Error(`GOOGLE_SEARCH_API_KEY is missing`);
  }
  if (cx == null) {
    throw new Error(`GOOGLE_SEARCH_ENGINE_ID is missing`);
  }
  const response = await fetch(
    "https://www.googleapis.com/customsearch/v1?" +
      new URLSearchParams({ key, cx, q, num: "5" }),
    { signal }
  );
  const json = await response.json();
  if (!GoogleSearchResponse.is(json)) {
    throw new Error("could not parse response from google search");
  }
  return (
    json.items?.map((r) => ({
      title: r.title ?? "<no title>",
      link: r.link ?? "<no link>",
      snippet: r.snippet ?? "<no snippet>",
    })) ?? []
  );
}
