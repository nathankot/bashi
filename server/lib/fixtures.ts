import type { Session } from "@lib/session.ts";
import type { CommandSet } from "@lib/command/types.ts";
import type { HandlerContext } from "$fresh/server.ts";
import builtinCommands from "@lib/command/builtinCommands.ts";

export const now = new Date(1671439270000);

export const handlerCtx: HandlerContext<{}> = {
  params: {},
  render: (data) => new Response(JSON.stringify(data)),
  renderNotFound: () => new Response("", { status: 404 }),
  state: {},
  localAddr: {
    transport: "tcp",
    hostname: "127.0.0.1",
    port: 80,
  },
  remoteAddr: {
    transport: "tcp",
    hostname: "127.0.0.1",
    port: 80,
  },
};

export const commandSet: CommandSet = {
  createCalendarEvent: {
    description: "create a calendar event on a certain date and time",
    args: [
      {
        name: "relative time supported by chrono-js",
        type: "string",
        parse: ["naturalLanguageDateTime"],
      },
      {
        name: "name",
        type: "string",
      },
    ],
    triggerTokens: ["calendar", "event"],
  },
  createReminder: {
    description: "create a reminder on a certain date and time",
    args: [
      {
        name: "relative time supported by chrono-js",
        type: "string",
        parse: ["naturalLanguageDateTime"],
      },
      { name: "name", type: "string" },
    ],
    triggerTokens: ["remind", "reminder"],
  },
  sendEmail: {
    description: `send an email`,
    args: [
      { name: "recipient", type: "string" },
      { name: "subject", type: "string" },
      { name: "contents", type: "string" },
    ],
    triggerTokens: ["email"],
  },
};

export const session: Session = {
  expiresAt: new Date(new Date().getTime() + 60000),
  sessionId: "a9fb6273-00ee-4e4c-9918-e87e1157ca31",
  accountNumber: "0000000000000000",
  configuration: {
    locale: "en-US",
    maxResponseTokens: 1000,
    bestOf: 1,
    enabledBuiltinCommands: Object.keys(builtinCommands) as any,
    timezoneUtcOffset: 0,
  },
  modelConfigurations: {
    "assist-000": {
      model: "assist-000",
      commands: commandSet,
    },
  },
};
