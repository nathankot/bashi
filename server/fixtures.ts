import { Session } from "@/types.ts";
import { HandlerContext } from "$fresh/server.ts";

export const handlerCtx: HandlerContext<{}> = {
  params: {},
  render: (data) => new Response(JSON.stringify(data)),
  renderNotFound: () => new Response("", { status: 404 }),
  state: {},
  localAddr: {
    transport: "tcp",
    hostname: "127.0.0.1",
    port: 80,
  },
  remoteAddr: {
    transport: "tcp",
    hostname: "127.0.0.1",
    port: 80,
  },
};

export const session: Session = {
  commands: [],
  expiresAt: new Date(new Date().getTime() + 60000),
  sessionId: "a9fb6273-00ee-4e4c-9918-e87e1157ca31",
};
