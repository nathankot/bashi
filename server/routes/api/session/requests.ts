import * as f from "fp-ts";

import { Handlers } from "$fresh/server.ts";

import { renderError, renderJSON, handleError } from "@lib/util.ts";

import { State as ApiState } from "@routes/api/_middleware.ts";
import { State } from "./_middleware.ts";
import { Input, Output, run, validateInput } from "@lib/models.ts";
import interceptors from "@lib/output_interceptors.ts";

export type PostRequestsRequest = Input;
export type PostRequestsResponse = Output;

export const handler: Handlers<Output, State & ApiState> = {
  async POST(req, ctx) {
    let json;
    try {
      json = await req.json();
    } catch {
      return renderError(400, "could not parse json");
    }

    const inputDecodeResult = Input.decode(json);
    if (!f.either.isRight(inputDecodeResult)) {
      return renderError(400, "malformed request");
    }

    const input: Input = inputDecodeResult.right as any;
    const modelDeps = { openai: ctx.state.clients.openai };
    const modelName = input.model;

    try {
      switch (modelName) {
        case "assist-davinci-003":
        case "noop":
          if (!validateInput(modelName, input)) {
            throw new Error("input could not be verified");
          }
          let output = await run(
            modelName,
            modelDeps,
            ctx.state.session,
            input
          );

          for (const interceptor of interceptors) {
            output = await interceptor(ctx.state.session, output);
          }

          return renderJSON(output);
        default:
          const exhaustiveCheck: never = modelName;
          throw new Error(`model ${exhaustiveCheck} not found`);
      }
    } catch (e) {
      return handleError(e);
    }
  },
};
