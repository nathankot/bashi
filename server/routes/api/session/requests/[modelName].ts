import * as f from "fp-ts";

import { Handlers } from "$fresh/server.ts";

import HTTPError from "@lib/http_error.ts";
import { renderError, renderJSON, handleError } from "@lib/util.ts";
import { wrap } from "@lib/log.ts";
import { defaultPolicy } from "@lib/faultHandling.ts";
import {
  AllInput,
  AllOutput,
  ModelName,
  InputFor,
  ModelDeps,
  run,
} from "@lib/models.ts";

import { State as ApiState } from "@routes/api/_middleware.ts";
import { State } from "../_middleware.ts";

export type Request = FormData | Omit<AllInput, "model">;

export const handler: Handlers<AllOutput, State & ApiState> = {
  async POST(req, ctx) {
    let log = ctx.state.log;

    const modelNameDecodeResult = ModelName.decode(ctx.params["modelName"]);
    if (!f.either.isRight(modelNameDecodeResult)) {
      return handleError(log, new HTTPError(`model name invalid`, 400));
    }
    const model: ModelName = modelNameDecodeResult.right as any;
    log = wrap({ model }, ctx.state.log);

    const modelDeps: Omit<ModelDeps, "signal"> = {
      log,
      openai: ctx.state.clients.openai,
      whisperEndpoint: ctx.state.clients.whisperEndpoint,
      session: ctx.state.session,
      now: ctx.state.now,
      faultHandlingPolicy: defaultPolicy,
    };

    try {
      switch (model) {
        case "whisper":
          const arrayBuffer = await req.arrayBuffer();
          if (arrayBuffer == null) {
            return renderError(400, "no audio found in the request body");
          }
          const arrayBufferInput: InputFor<typeof model> = {
            model,
            arrayBuffer,
          };

          return renderJSON(await run(modelDeps, model, arrayBufferInput));

        case "assist-000":
        case "translate-davinci-003":
        case "code-000":
        case "noop":
          // JSON request
          let json;
          try {
            json = await req.json();
          } catch {
            return renderError(400, "could not parse json");
          }

          const inputDecodeResult = AllInput.decode({ model, ...json });
          if (!f.either.isRight(inputDecodeResult)) {
            return renderError(400, "malformed request");
          }

          const requestInput: InputFor<typeof model> =
            inputDecodeResult.right as any;

          if (
            "request" in requestInput &&
            requestInput.request.trim().length === 0
          ) {
            return renderError(400, "request must not be empty");
          }

          return renderJSON(await run(modelDeps, model, requestInput));

        default:
          const exhaustiveCheck: never = model;
          throw new Error(`model ${exhaustiveCheck} not found`);
      }
    } catch (e) {
      log("error", e);
      return handleError(log, e);
    }
  },
};
