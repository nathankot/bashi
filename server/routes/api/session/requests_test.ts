import { assertSnapshot } from "std/testing/snapshot.ts";
import { stub } from "std/testing/mock.ts";

import * as fixtures from "@lib/fixtures.ts";
import * as clients from "@lib/clients.ts";
import { handler } from "./requests.ts";

const nowFixture = new Date();

for (const test of [
  {
    description: "success",
    request: `{ "model": "assist-davinci-003", "request": "a mock request" }`,
    session: fixtures.session,
    openAiFn: async () =>
      ({
        data: {
          choices: [
            {
              text: "fake response",
            },
          ],
        },
      } as any),
  },
  {
    description: "openai errors",
    request: `{ "model": "assist-davinci-003", "request": "a mock request" }`,
    session: fixtures.session,
    openAiFn: async () => {
      throw new Error("mock error");
    },
  },
  // too many tokens
  // rate limited (or maybe handled by middleware?)
]) {
  Deno.test("POST /api/session/requests: " + test.description, async (t) => {
    const openAiStub = stub(clients.openai, "createCompletion", test.openAiFn);

    try {
      const enc = new TextEncoder();
      const req = new Request("http://localhost/", {
        method: "POST",
        body: enc.encode(test.request),
      });
      req.headers.set("Content-Type", "application/json");
      const response = await handler.POST!(req, {
        ...fixtures.handlerCtx,
        state: {
          now: nowFixture,
          clients,
          session: test.session,
        },
      });
      const responseJson = await response.json();
      await assertSnapshot(t, responseJson);
    } finally {
      openAiStub.restore();
    }
  });
}
