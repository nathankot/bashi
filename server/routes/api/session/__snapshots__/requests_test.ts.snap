export const snapshot = {};

snapshot[`POST /api/session/requests: success 1`] = `
{
  functionCalls: [
    {
      args: [
        "mock response",
      ],
      name: "respond",
      type: "parsed",
    },
    {
      args: [
        123,
      ],
      invalid_reason: "unknown_function",
      name: "unknownCall",
      type: "parsed_but_invalid",
    },
    {
      error: "undefined: Unable to consume token: <END-OF-FILE>",
      line: "unparseable(",
      type: "parse_error",
    },
    {
      args: [
        "America/New_York",
      ],
      name: "time",
      type: "parsed",
    },
  ],
  model: "assist-davinci-003",
  request: "a mock request",
}
`;

snapshot[`POST /api/session/requests: openai errors 1`] = `
{
  error: "internal server error",
}
`;

snapshot[`POST /api/session/requests: noop model works 1`] = `
{
  model: "noop",
}
`;
